packages/datastore/src/datastore/datastore.ts

```diff
@@ -502,6 +520,73 @@ const createModelClass = <T extends PersistentModel>(
 
 	Object.defineProperty(clazz, 'name', { value: modelDefinition.name });
 
+	for (const field in modelDefinition.fields) {
+		if (!isFieldAssociation(modelDefinition, field)) {
+			continue;
+		}
+
+		const {
+			type,
+			association: { targetName },
+		} = modelDefinition.fields[field];
+		const relatedModelName = type['model'];
+
+		Object.defineProperty(clazz.prototype, modelDefinition.fields[field].name, {
+			set(model: PersistentModel) {
+				if (!model || !model.id) return;
+				if (model.hasOwnProperty('_version')) {
+					const modelConstructor = Object.getPrototypeOf(model || {})
+						.constructor as PersistentModelConstructor<T>;
+
+					if (!isValidModelConstructor(modelConstructor)) {
+						const msg = `Value passed to ${modelDefinition.name}.${field} is not a valid instance of a model`;
+						logger.error(msg, { model });
+
+						throw new Error(msg);
+					}
+
+					if (
+						modelConstructor.name.toLowerCase() !==
+						relatedModelName.toLowerCase()
+					) {
+						const msg = `Value passed to ${modelDefinition.name}.${field} is not an instance of ${relatedModelName}`;
+						logger.error(msg, { model });
+
+						throw new Error(msg);
+					}
+				}
+				if (targetName) {
+					this[targetName] = model.id;
+				}
+			},
+			async get() {
+				if (!memos.has(this)) {
+					memos.set(this, {});
+				}
+				const instanceMemos = memos.get(this);
```

Comment:
Can simplify this a bit, since you're doing a `set` at the bottom of the method anyways:

```suggestion
				const instanceMemos = memos.get(this) || {};
```